src/odemis/odemisd/main.py

```diff
@@ -90,40 +92,54 @@ def __init__(self, model_file, settings_file, create_sub_containers=False,
             logging.exception("When instantiating file %s", self._model.name)
             raise IOError("Unexpected error at instantiation")
 
-    def _on_va_change(self, _):
+    def _observe_persistent_va(self, comp, prop_name):
         """
-        Callback to VA change, calls write_persistent_data
+        Listen to changes in persistent va, update the value of instantiator.persistent_props
+        and write data to file.
+        comp (HwComponent): component
+        prop_name (str): name of va
         """
-        self._write_persistent_data()
+        def on_va_change(value, comp=comp.name, prop=prop_name):
+            # Update persistent properties in instantiator
+            if comp not in self._instantiator.persistent_props:
+                self._instantiator.persistent_props[comp] = {}
+            self._instantiator.persistent_props[comp][prop] = value
+            self._write_persistent_data()
+
+        va = getattr(comp, prop_name)
+        va.subscribe(on_va_change)
+        self.persistent_listeners.append(on_va_change)
 
     def _write_persistent_data(self):
         """
         Write values for all persistent properties and metadata to the settings file.
         """
         data = {}
-        for comp in self._instantiator.components:
-            props, mds = self._instantiator.get_persistent(comp)
-            # Initialize internal dict
-            comp_data = {}
-            if props:
-                comp_data['properties'] = {}
-            if mds:
-                comp_data['metadata'] = {}
-
-            for prop in props:
-                comp_data['properties'][prop] = getattr(comp, prop).value
-            for md in mds:
-                fullname = "MD_" + md
-                comp_data['metadata'][md] = comp.getMetadata()[getattr(model, fullname)]
-
-            # Only store components that actually contain persistent data
-            if comp_data:
-                data[comp.name] = comp_data
+        for comp, prop_dict in self._instantiator.persistent_props.items():
```

Comment:
I think it'd be simpler if you just kept the whole structure in the right format in memory in the backend.
Then this function would be just 3 lines.

To import the original values from the instantiator, either use this code in the init, or just let the instantiator directly provide the whole structure as-is, and as read from the YAML file.